lib/Parse/ParsePattern.cpp

```diff
@@ -888,6 +879,56 @@ void Parser::parseAsyncThrows(
   }
 }
 
+/// Parse a type in parentheses after a `throws` or `rethrows`
+///
+ParserResult<TypeRepr> Parser::parseThrowsType() {
+  // look for the left paren for throws (_type_)
+  BacktrackingScope backtrack(*this);
+  
+  ParserResult<TypeRepr> throwsTypeRepr;
+  
+  if (Tok.is(tok::l_paren)) {
+    consumeToken();
+    throwsTypeRepr = parseTypeIdentifier();
+    if (!throwsTypeRepr.isParseError()) {
+      if (Tok.is(tok::r_paren)) {
+        consumeToken();
```

Comment:
here we need a consume with a diagnostic, that a r_paren is missing

Reply:
In the `else` branch of this one we should generate a diagnostic and, I guess, just leave the cursor where it is. There are other cases after a left-paren that I need to break out, probably:

1. immediate right-paren
3. type identifier followed by something that's not a right paren (the case here). 
2. unparseable non-type identifier stuff before right-paren (probably diagnostic, skip the left paren, and let it try to recover. 
4. some other possibilities I have not thought of yet.